### ISSUE TITLE
New Feature request about GGUI : Mesh Instancing

### ISSUE THREAD

**Issue by Morcki** [taichi-dev/taichi#5513]
**Feature Description**
> We would like to add Mesh Instancing in GGUI part to support users to draw lots of the same mesh with different attributes by one function.

**Here are some details of our designs now**

```python
def mesh_instance(self,
                      vertices,
                      indices,
                      normals,
                      color,
                      per_vertex_color,
                      two_sided,
                      translation_vector_list = None, # translation of each mesh instance
                      scale_vector_list = None, # scale of each mesh instance
                      rotation_vector_list = None, # rotation of each mesh instance
                      vertex_offset: int = 0,
                      vertex_count: int = None,
                      index_offset: int = 0,
                      index_count: int = None):
```
1. `translation_vector_list` is a ti.Vector.field with 3 dimension and (N,) shape. You can create your translation vector of each instance into the field. After you change the translation of some instances, you should update this vector field and call this function.
2. `scale_vector_list` is also a ti.Vector.field with 3 dimension and (N,) shape. And usage is the same with above description.
3. `rotation_vector_list` is a ti.Vector.field with 3 dimension and (N,) shape. Here rotation means eular angle(X, Y, Z), After you pass the value into this function. On cpp part, we will creates a 3D 3 * 3 rotation matrix from euler angles (with rotation order Y * X * Z), and then convert rotation matrix to quaternion.

Note that the count of mesh instances is N which is the minimum of  the shape of 3 vector lists. 
+ The advantage of this design is that you can update any mesh instance transformation easily (just get the index of the instance, and update its translation, scale and rotation, then leave it alone).
+ The disadvantage is that you have to manage 3 vector field by yourself, and rotation order is fixed.

**Some other suggestions?**

For example, you may want to pass a quaternion, or you want to decide the roation order by yourself. Moreover, you may even want to pass the transformation matrix directly (create a transformation class to handle translation, scalar and rotation). 
Let's talk deeply into that, and we hope to provide you with the most convenient function.

**Welcom and give us any wild ideas about what you would like this feature to be**

**Comment by bobcao3** [comment]
I would prefer we directly use a transformation matrix list and create utility functions to generate these matrices easily.

**Comment by PENGUINLIONG** [comment]
eular angles could be a problem for animated rotation. so if we dont consume a list of prebuilt matrices, i think quaternions would be better.

**Comment by Morcki** [comment]
> I would prefer we directly use a transformation matrix list and create utility functions to generate these matrices easily.

Yes, I have considered about that. So this solution means that we shoud focus on the Transformation Class. Maybe we can give more details about the API of this class?

**Comment by Morcki** [comment]
> eular angles could be a problem for animated rotation. so if we dont consume a list of prebuilt matrices, i think quaternions would be better.

So your opnion is that users should pass quaternion vector list. And we should support some build-in function in Matrix.field or Vector.field to convert the eular angle to quaternion in user-customized order?

**Comment by Morcki** [comment]
Hi, as mentioned above, I think that passing the transformation matrix directly is convenient for most people at now. So the API at present will be as below:
```python
def mesh_instance(self,
                      vertices,
                      indices,
                      normals,
                      color,
                      per_vertex_color,
                      two_sided,
                      transform_matrix_list : ti.field((4, 4), ti.f32, shape = N),
                      vertex_offset: int = 0,
                      vertex_count: int = None,
                      index_offset: int = 0,
                      index_count: int = None):
```
I will support some built-in API for python users to convert translaton vector, scale vector, rotation vector to Matrix, and user can use these APIs to build a transformation matrix by themself.

And I will release a pre-version for internal test and collect feedback. 
If you have any better design pattern, please share your ideas and list details here.

**Comment by Morcki** [comment]
New added built-in APIs for constructing a transform matrix are as below, details please refer to the notes of each API, or refer to #5827  

- Translation
  - ti.math.translate(dx, dy, dz)
- Scale
  - ti.math.scale(sx, sy, sz)
- Rotation
  - ti.math.rot_by_axis(ang, axis)
  - ti.math.rot_yaw_pitch_roll(yaw, pitch, roll)
  - ti.math.rotation3d(ang_x, ang_y, ang_z)

::: Note that all the rotations ars based on counter-clockwise
